lib/token.h

```diff
@@ -187,6 +187,30 @@ class CPPCHECKLIB Token {
     explicit Token(TokensFrontBack *tokensFrontBack = nullptr);
     ~Token();
 
+    struct TokenEnumeration
```

Comment:
@danmar regarding the question of whether this should be in Token, mTokensFrontBack is a private member of Token so if this class isn't part of Token I would have to expose mTokensFrontBack somehow.

Reply:
This should probably be templated so it can iterate over both `Token* and `const Token*`(probably as `template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)> struct TokenEnumeration` so the template errors are clearer).

Also, this is considered a range in C++. It seems like `TokenRange` would be a better name, but its up to you.

Reply:
I like the name `TokenRange`.

I am not sure if we need `Token*`.. I spontanously feel it would not be extremely useful. If we iterate the non-const tokens and modify the list somehow then a range loop might not be a good idea. Probably it's not very useful in Tokenizer. In createAst I don't think we could use this neither. Would that be useful in ValueFlow?

Reply:
Even if we don't want a non-const version (which I also couldn't think of a use for) it might be worth explicitly disabling for improved compiler diagnostics as suggested. At that point it might be simpler to implement rather than not.

Reply:
unfortunately the example given for the REQUIRES macro prevents forward declaration of the type which I think is a problem

Reply:
> Would that be useful in ValueFlow?

> If we iterate the non-const tokens and modify the list somehow then a range loop might not be a good idea. 

We should probably extend the `invalidContainerLoop` to detect this issue with tokens.

Yes, its quite common in ValueFlow to use `Token*` because we need to modify the values in the token. However, we dont modify the tokens order in ValueFlow.

Of course, it would be nice if we could get rid of this, as it could simplify a lot of things such as extra overloads for both const and non-const in astutils and reusing the analyzers outside of valueflow. One way to do this is to make the values a pointer and then in valueflow use a map from the token to values pointer(ie `std::map<const Token*, std::shared_ptr<std::list<ValueFlow::Value>>>`) this way valueflow can update the values using `const Token*`. 

Although, for the short-term I dont see that happening, so it seems it would be good to have both.

Reply:
> unfortunately the example given for the REQUIRES macro prevents forward declaration of the type which I think is a problem

We should template a base class and then inherit it for const and non-const versions:

```cpp
template<class T, REQUIRES("T must be a Token class", std::is_convertible<T*, const Token*>)>
struct TokenRangeBase {
    // ...
};

struct ConstTokenRange : TokenRangeBase<const Token> {
    using TokenRangeBase<const Token>::TokenRangeBase<const Token>;
};

struct TokenRange : TokenRangeBase<Token> {
    using TokenRangeBase<Token>::TokenRangeBase<Token>;
};
```

That should simplify forward declarations.

Reply:
This is no longer true since I'm using nullptr rather than mTokensFrontBack

Reply:
good idea